packages/apps/dashboard/server/src/modules/details/details.service.ts

```diff
@@ -148,40 +153,88 @@ export class DetailsService {
     return result;
   }
 
-  public async getLeadersByChainId(
+  public async getLeaders(
     chainId?: ChainId,
-    take?: number,
+    orderBy?: LeadersOrderBy,
+    orderDirection?: OrderDirection,
+    first?: number,
+    skip?: number,
   ): Promise<LeaderDto[]> {
-    const chainIds = !chainId
-      ? await this.networkConfig.getAvailableNetworks()
-      : [chainId];
+    const chainIds = chainId
+      ? [chainId]
+      : await this.networkConfig.getAvailableNetworks();
 
-    let allLeadersData: ILeader[] = [];
-    for (const id of chainIds) {
-      const leadersData = await OperatorUtils.getLeaders({ chainId: id });
-      allLeadersData = allLeadersData.concat(
-        leadersData.filter((leader) => leader.amountStaked > 0 && leader.role),
+    const results = await Promise.all(
+      chainIds.map(async (id) => {
+        const filter = this.createLeadersFilter(
+          id,
+          orderBy,
+          orderDirection,
+          first,
+          skip,
+        );
+        const [leaders, reputations] = await Promise.all([
+          OperatorUtils.getLeaders(filter),
+          this.fetchReputations(id),
+        ]);
+        return { leaders, reputations };
+      }),
+    );
+
+    const leadersData = results.flatMap(({ leaders }) => leaders);
+    const reputations = results.flatMap(({ reputations }) => reputations);
+
+    const leaders = leadersData
+      .filter((leader) => leader.role)
+      .map((leader) =>
+        plainToInstance(LeaderDto, leader, { excludeExtraneousValues: true }),
       );
-    }
 
-    allLeadersData.sort((a, b) =>
-      BigInt(a.amountStaked) >= BigInt(b.amountStaked) ? -1 : 1,
-    );
-    if (take && take > 0) {
-      allLeadersData = allLeadersData.slice(0, take);
+    this.assignReputationsToLeaders(leaders, reputations);
+
+    if (orderBy === LeadersOrderBy.REPUTATION) {
```

Comment:
In case if we need to get list of leaders sorted by reputation we need to:
- first: get all reputations sorted asc/desc
- take the page of N addresses from sorted reputations
- retrieve leaders details for these N addresses
- return the results

Retrieving leaders by addresses can be done in 3 ways:
1. by using `OperatorUtils.getLeader` for each address
2. by adding `addresses` filter to `OperatorUtils.getLeaders`
3. implementing new util like `OperatorUtils.getLeadersByIds`

The first option is not effective and it might be expensive to send many queries to subgraph.
The second one seems a bit dirty because it becomes a "super-method", so I would do it as a 3rd one, where we can also limit the number of `ids` to get to some reasonable value (e.g. based on GQL max query size) and that might be less than max value of `1000` allowed for page

As an alternative and as option to simplify this PR - we can remove the "feature" of sorting by "reputation"

wdyt, @portuu3 @eugenvoronov ?

Reply:
It is a good point, but I think we need reputation ordering.
In any case max items in GQL is 1000, I don't think we will have 1000 operators at least for now, so it could be addresses in one request to subgraph and other to reputation oracle and cross the data.

Reply:
@dnechay Before creating this PR, I attempted to implement option 2. However, I found that when requesting a list of leaders by addresses from the subgraph in a specific order, the subgraph does not preserve the order and disrupts it.

Copy of the message to Adrian: 
> I tested a slightly more optimized approach for filtering leaders. The idea was to first request data filtered by reputation points from the Reputation Oracle, and then query the SDK for the leaders by their addresses, fetching only the necessary ones in the already defined order. Unfortunately, when selecting data from the subgraph using addresses, the order wasn’t preserved (because subgraph didn't save the order of records), so this approach ended up being ineffective.
I’ve reverted to the original plan: retrieving the full list of leaders first from subgraph and then sorting them on the dashboard side.
[4:07](https://humanprotocol.slack.com/archives/D04AE6252QL/p1733922466909459)
This is how I wanted to get them from subgraph, i've passed the list of addresses of leaders received from Reputation Oracle.
export const GET_LEADERS_QUERY = (filter: ILeadersFilter) => {
  const { role,, addresses } = filter;

  const WHERE_CLAUSE = `
    where: {
      ${addresses ? `address_in: $addresses` : ''}
      ${role ? `role: $role` : ''}
    }
  `;

  return gql`
    query getLeaders(
      $addresses: [String]
      $role: String
      $first: Int
      $skip: Int
      $orderBy: String
      $orderDirection: String
    ) {
      leaders(
        ${WHERE_CLAUSE}
        first: $first
        skip: $skip
        orderBy: $orderBy
        orderDirection: $orderDirection
      ) {
        ...LeaderFields
      }
    }
    ${LEADER_FRAGMENT}
  `;
};

Reply:
This may be the best and easiest way we can do this:

> It is a good point, but I think we need reputation ordering. In any case max items in GQL is 1000, I don't think we will have 1000 operators at least for now, so it could be addresses in one request to subgraph and other to reputation oracle and cross the data. @dnechay